check.go

```diff
@@ -0,0 +1,1086 @@
+// Errors are returned if an attempt to delete an item fails.
+func (s *store) Repair(report CheckReport, options *RepairOptions) []error {
+	if options == nil {
+		options = RepairEverything()
+	}
+	var errs []error
+	// Just delete damaged containers.
+	if options.RemoveContainers {
+		for id := range report.Containers {
+			err := s.DeleteContainer(id)
+			if err != nil && !errors.Is(err, ErrContainerUnknown) {
+				err := fmt.Errorf("deleting container %s: %w", id, err)
+				errs = append(errs, err)
+			}
+		}
+	}
+	// Now delete damaged images.  Note which layers were removed as part of removing those images.
+	deletedLayers := make(map[string]struct{})
+	for id := range report.Images {
+		layers, err := s.DeleteImage(id, true)
+		if err != nil {
+			if !errors.Is(err, ErrImageUnknown) && !errors.Is(err, ErrLayerUnknown) {
+				err := fmt.Errorf("deleting image %s: %w", id, err)
+				errs = append(errs, err)
+			}
+		} else {
+			for _, layer := range layers {
+				logrus.Debugf("deleted layer %s", layer)
+				deletedLayers[layer] = struct{}{}
+			}
+			logrus.Debugf("deleted image %s", id)
+		}
+	}
+	// Build a list of the layers that we need to remove, sorted with parents of layers before
+	// layers that they are parents of.
+	layersToDelete := make([]string, 0, len(report.Layers))
+	for id := range report.Layers {
+		layersToDelete = append(layersToDelete, id)
+	}
+	depth := func(id string) int {
+		d := 0
+		parent := report.layerParentsByLayerID[id]
+		for parent != "" {
+			d++
+			parent = report.layerParentsByLayerID[parent]
+		}
+		return d
+	}
+	isUnaccounted := func(errs []error) bool {
+		for _, err := range errs {
+			if errors.Is(err, ErrLayerUnaccounted) {
+				return true
+			}
+		}
+		return false
+	}
+	sort.Slice(layersToDelete, func(i, j int) bool {
+		// we've not heard of either of them, so remove them in the order the driver suggested
+		if isUnaccounted(report.Layers[layersToDelete[i]]) &&
+			isUnaccounted(report.Layers[layersToDelete[j]]) &&
+			report.layerOrder[layersToDelete[i]] != 0 && report.layerOrder[layersToDelete[j]] != 0 {
+			return report.layerOrder[layersToDelete[i]] < report.layerOrder[layersToDelete[j]]
+		}
+		// always delete the one we've heard of first
+		if isUnaccounted(report.Layers[layersToDelete[i]]) && !isUnaccounted(report.Layers[layersToDelete[j]]) {
+			return false
+		}
+		// always delete the one we've heard of first
+		if !isUnaccounted(report.Layers[layersToDelete[i]]) && isUnaccounted(report.Layers[layersToDelete[j]]) {
+			return true
+		}
+		// we've heard of both of them; the one that's on the end of a longer chain goes first
+		return depth(layersToDelete[i]) > depth(layersToDelete[j]) // closer-to-a-notional-base layers get removed later
+	})
+	// Now delete the layers that haven't been removed along with images.
+	for _, id := range layersToDelete {
+		if _, ok := deletedLayers[id]; ok {
+			continue
+		}
+		for _, reportedErr := range report.Layers[id] {
```

Comment:
This could try to remove the same layer multiple times, is that intentional?

I’d expect a single attempt, at the driver or higher level depending on `isUnaccounted`.